lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,950 @@
+
+  // We'll want the CodingKeys enum for this type.
+  NominalTypeDecl *codingKeysDecl;
+  bool derived;
+  std::tie(codingKeysDecl, derived) = getOrSynthesizeCodingKeysDecl(tc,
+                                                                    typeDecl,
+                                                                    typeDecl);
+
+  // We should have bailed already if:
+  // a) The type does not have CodingKeys
+  assert(codingKeysDecl && "Missing CodingKeys decl.");
+  // b) We had to derive it right here
+  assert(!derived && "Had to derive CodingKeys type too late.");
+  // c) The type is not an enum
+  auto codingKeysEnum = cast<EnumDecl>(codingKeysDecl);
+
+  SmallVector<ASTNode, 5> statements;
+
+  // Generate a reference to containerExpr ahead of time in case there are no
+  // properties to encode or decode, but the type is a class which inherits from
+  // something Codable and needs to encode super.
+
+  // let container : KeyedEncodingContainer<CodingKeys>
+  auto codingKeysType = codingKeysEnum->getDeclaredType();
+  auto containerDecl = createKeyedContainer(C, funcDC,
+                                            C.Id_KeyedEncodingContainer,
+                                            codingKeysType);
+
+  auto containerExpr = new (C) DeclRefExpr(ConcreteDeclRef(containerDecl),
+                                           DeclNameLoc(), /*Implicit=*/true,
+                                           AccessSemantics::DirectToStorage);
+
+  auto enumElements = codingKeysEnum->getAllElements();
+  if (!enumElements.empty()) {
+    // Need to generate
+    //   `let container = encoder.container(keyedBy: CodingKeys.self)`
+    // `let container` (containerExpr) is generated above.
+
+    // encoder
+    auto encoderParam = encodeDecl->getParameterList(1)->get(0);
+    auto encoderExpr = new (C) DeclRefExpr(ConcreteDeclRef(encoderParam),
+                                           DeclNameLoc(), /*Implicit=*/true);
+
+    // Bound encoder.container(keyedBy: CodingKeys.self) call
+    auto containerType = containerDecl->getInterfaceType();
+    auto callExpr = createContainerKeyedByCall(C, funcDC, encoderExpr,
+                                               containerType, codingKeysEnum);
+
+    // Full `let container = encoder.container(keyedBy: CodingKeys.self)`
+    // binding.
+    auto containerPattern = new (C) NamedPattern(containerDecl,
+                                                 /*implicit=*/true);
+    auto bindingDecl = PatternBindingDecl::create(C, SourceLoc(),
+                                                  StaticSpellingKind::None,
+                                                  SourceLoc(), containerPattern,
+                                                  callExpr, funcDC);
+    statements.push_back(bindingDecl);
+    statements.push_back(containerDecl);
+
+    // Now need to generate `try container.encode(x, forKey: .x)` for all
+    // existing properties.
+    for (auto elt : enumElements) {
```

Comment:
Is this a pointer? Or a reference? In these cases it is good to put in the '*' or the '&' so people down the line are able to understand the ownership semantics of the for each. We have had several cases of people assuming an auto was a & when it was really a *& resulting in miscompiles.